lib/cache/sqlite-cache-store.js

```diff
@@ -0,0 +1,449 @@
+    /**
+     * @type {Buffer[] | null}
+     */
+    let body = key.method !== 'HEAD' ? [] : null
+    const maxEntrySize = this.#maxEntrySize
+    const findValue = this.#findValue.bind(this)
+    const updateValueQuery = this.#updateValueQuery
+    const insertValueQuery = this.#insertValueQuery
+
+    this.prune()
+
+    const writable = new Writable({
+      write (chunk, encoding, callback) {
+        if (typeof chunk === 'string') {
+          chunk = Buffer.from(chunk, encoding)
+        }
+
+        currentSize += chunk.byteLength
+
+        if (body) {
+          if (currentSize >= maxEntrySize) {
+            body = null
+            this.end()
+            return callback()
+          }
+
+          body.push(chunk)
+        }
+
+        callback()
+      },
+      final (callback) {
+        if (body === null) {
+          return callback()
+        }
+
+        /**
+         * @type {SqliteStoreValue | undefined}
+         */
+        const existingValue = findValue(key, true)
+        if (existingValue) {
+          // Updating an existing response, let's delete it
+          updateValueQuery.run(
+            JSON.stringify(stringifyBufferArray(body)),
+            value.deleteAt,
+            value.statusCode,
+            value.statusMessage,
+            value.rawHeaders ? JSON.stringify(stringifyBufferArray(value.rawHeaders)) : null,
+            value.etag,
+            value.cachedAt,
+            value.staleAt,
+            value.deleteAt,
+            existingValue.id
+          )
+        } else {
+          // New response, let's insert it
+          insertValueQuery.run(
+            url,
+            key.method,
+            JSON.stringify(stringifyBufferArray(body)),
+            value.deleteAt,
+            value.statusCode,
+            value.statusMessage,
+            value.rawHeaders ? JSON.stringify(stringifyBufferArray(value.rawHeaders)) : null,
+            value.etag ? value.etag : null,
+            value.vary ? JSON.stringify(value.vary) : null,
+            value.cachedAt,
+            value.staleAt,
+            value.deleteAt
+          )
+        }
+
+        callback()
+      }
+    })
+
+    return writable
+  }
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheKey} key
+   */
+  delete (key) {
+    if (typeof key !== 'object') {
+      throw new TypeError(`expected key to be object, got ${typeof key}`)
+    }
+
+    this.#deleteByUrlQuery.run(this.#makeValueUrl(key))
+  }
+
+  /**
+   * This method is called to prune the cache when it exceeds the maximum number
+   * of entries. It removes half the entries in the cache, ordering them the oldest.
+   *
+   * @returns {Number} The number of entries removed
+   */
+  prune () {
+    const total = this.size()
+
+    if (total <= this.#maxDbEntries) {
+      return
+    }
+
+    const oldestCachedAt = this.#getOldCacheAtQuery.get(total - 1)?.cachedAt
+
+    const res = this.#deleteOldCacheAtQuery.run(oldestCachedAt)
```

Comment:
Can't you do this in a single query?

```delete from XXX where id IN (SELECT id from XXX order by cachedAt DESC limit 20)```

Reply:
Also I would prune at least 10% of elements to we do not have to run this so often...

Reply:
I would also move `this.#deleteExpiredValuesQuery.run(now)` to here and have it first try that and see if it's sufficient and then do a drop oldest.